BuildMasterAutomation/Functions/Get-BMDeployment.ps1

```diff
@@ -30,29 +53,88 @@ function Get-BMDeployment
+        [Parameter(ParameterSetName='ByFilter')]
+        [Object] $Release,
+
+        # The build to get deployments for. You can pass an build id, build name, build number, or build object.
+        [Parameter(ParameterSetName='ByFilter')]
+        [Object] $Build,
+
+        # The environment to get deployments for. You can pass an environment id, environment name, or environment object.
+        [Parameter(ParameterSetName='ByFilter')]
+        [Object] $Environment,
+
+        # The name of the pipeline to get deployments for.
+        [Parameter(ParameterSetName='ByFilter')]
+        [String] $Pipeline,
+
+        # The name of the pipeline stage to get deployments for.
+        [Parameter(ParameterSetName='ByFilter')]
+        [String] $Stage,
+
+        # The status of the deployments to get. Accepted values are 'pending', 'executing', 'succeeded', 'warned', or 'failed'
+        [Parameter(ParameterSetName='ByFilter')]
+        [ValidateSet('pending', 'executing', 'succeeded', 'warned', 'failed')]
+        [String] $Status
     )
 
     process
     {
+
         Set-StrictMode -Version 'Latest'
         Use-CallerPreference -Cmdlet $PSCmdlet -SessionState $ExecutionContext.SessionState
         $WhatIfPreference = $false
 
         $parameter =
             @{ } |
-            Add-BMObjectParameter -Name 'deployment' -Value $Deployment -PassThru
+            Add-BMObjectParameter -Name 'deployment' -Value $Deployment -PassThru |
+            Add-BMObjectParameter -Name 'application' -Value $Application -PassThru |
+            Add-BMObjectParameter -Name 'environment' -Value $Environment -PassThru |
+            Add-BMObjectParameter -Name 'release' -Value $Release -PassThru
+
+        if($Build)
+        {
+            if ($Build -is [string] -and -not [Int64]::TryParse($Build, [ref] $Build))
+            {
+                $parameter['buildNumber'] = $Build
+            }
+            else
+            {
+                $parameter = $parameter | Add-BMObjectParameter -Name 'build' -Value $Build -PassThru
+            }
+        }
+
+        if($Pipeline)
+        {
+            $parameter['pipelineName'] = $Pipeline
+        }
+        if($Stage)
+        {
+            $parameter['pipelineStageName'] = $Stage
+        }
+        if($Status)
+        {
+            $parameter['status'] = $Status
+        }
 
         $deployments = @()
         Invoke-BMRestMethod -Session $Session -Name 'releases/builds/deployments' -Parameter $parameter -Method Post |
             Tee-Object -Variable 'deployments' |
             Write-Output
 
-        if (-not $deployments)
+        if(-not $deployments)
         {
-            $msg = "Unable to get deployment ""$($Deployment | Get-BMObjectName)"" because it does not exist."
+            $params = ($parameter.Keys | ForEach-Object { "$($_): $($parameter[$_])"}) -join ', '
+            $msg = "Unable to get deployment with parameters ""$($params)"" because it does not exist."

```

Comment:
Rework so that the error message aligns with how the user called the function:

```powershell
if (-not $deployments)
{
    if ($PSCmdlet.ParameterSetName -eq 'ById')
    {
        $msg = "Unable to get deployment ""$($Deployment | Get-BMObjectName)"" because it does not exist."
    }
    else
    {
        $params = ($parameter.Keys | ForEach-Object { "$($_) = $($parameter[$_])"}) -join ', '
        $msg = "Unable to find any deployments that match: ""$($params)""."
    }

    Write-Error -Message $msg -ErrorAction $ErrorActionPreference
}
```